typing/parmatch.ml

```diff
@@ -358,93 +506,61 @@ let get_constructor_type_path ty tenv =
-  match p1.pat_desc, p2.pat_desc with
-  | Tpat_construct(_, c1, _), Tpat_construct(_, c2, _) ->
+let simple_match d1 d2 =
+  match Pattern_head.desc d1, Pattern_head.desc d2 with
+  | Construct c1, Construct c2 ->
       Types.equal_tag c1.cstr_tag c2.cstr_tag
-  | Tpat_variant(l1, _, _), Tpat_variant(l2, _, _) ->
-      l1 = l2
-  | Tpat_constant(c1), Tpat_constant(c2) -> const_compare c1 c2 = 0
-  | Tpat_lazy _, Tpat_lazy _ -> true
-  | Tpat_record _ , Tpat_record _ -> true
-  | Tpat_tuple p1s, Tpat_tuple p2s
-  | Tpat_array p1s, Tpat_array p2s -> List.length p1s = List.length p2s
-  | _, (Tpat_any | Tpat_var(_)) -> true
+  | Variant { tag = t1; _ }, Variant { tag = t2 } ->
+      t1 = t2
+  | Constant c1, Constant c2 -> const_compare c1 c2 = 0
+  | Lazy, Lazy -> true
+  | Record _, Record _ -> true
+  | Tuple len1, Tuple len2
+  | Array len1, Array len2 -> len1 = len2
+  | _, Any -> true
   | _, _ -> false
 
 
 
-
 (* extract record fields as a whole *)
-let record_arg p = match p.pat_desc with
-| Tpat_any -> []
-| Tpat_record (args,_) -> args
+let record_arg p = match Pattern_head.desc p with
+| Any -> []
+| Record args -> args
 | _ -> fatal_error "Parmatch.as_record"
 
 
 (* Raise Not_found when pos is not present in arg *)
 let get_field pos arg =
-  let _,_, p = List.find (fun (_,lbl,_) -> pos = lbl.lbl_pos) arg in
+  let _, p = List.find (fun (lbl,_) -> pos = lbl.lbl_pos) arg in
   p
 
-let extract_fields omegas arg =
-  List.map
-    (fun (_,lbl,_) ->
-      try
-        get_field lbl.lbl_pos arg
-      with Not_found -> omega)
-    omegas
+let extract_fields lbls arg =
+  List.map (fun lbl ->
+      try get_field lbl.lbl_pos arg
+      with Not_found -> omega
+  ) lbls
 
 (* Build argument list when p2 >= p1, where p1 is a simple pattern *)
-let rec simple_match_args p1 p2 = match p2.pat_desc with
-| Tpat_alias (p2,_,_) -> simple_match_args p1 p2
-| Tpat_construct(_, _, args) -> args
-| Tpat_variant(_, Some arg, _) -> [arg]
-| Tpat_tuple(args)  -> args
-| Tpat_record(args,_) ->  extract_fields (record_arg p1) args
-| Tpat_array(args) -> args
-| Tpat_lazy arg -> [arg]
-| (Tpat_any | Tpat_var(_)) ->
-    begin match p1.pat_desc with
-      Tpat_construct(_, _,args) -> omega_list args
-    | Tpat_variant(_, Some _, _) -> [omega]
-    | Tpat_tuple(args) -> omega_list args
-    | Tpat_record(args,_) ->  omega_list args
-    | Tpat_array(args) ->  omega_list args
-    | Tpat_lazy _ -> [omega]
-    | _ -> []
+let simple_match_args d1 d2 args = match Pattern_head.desc d2 with
+| Construct _
+| Variant _
+| Tuple _
+| Array _
+| Lazy -> args
+| Record lbls ->  extract_fields (record_arg d1) (List.combine lbls args)
+| Any ->
+    begin match Pattern_head.desc d1 with
+    | Construct cstr -> omegas cstr.cstr_arity
+    | Variant { has_arg = true }
+    | Lazy -> [omega]
+    | Record lbls ->  omega_list lbls
+    | Array len
+    | Tuple len -> omegas len
+    | Variant { has_arg = false }
+    | Any
+    | Constant _ -> []
```

Comment:
This inner match looks a lot like a smaller sibling of `Pattern_head.to_omega_pattern`. Maybe, it could be defined in `Pattern_head` too?

Reply:
I'm not keen, `to_omega_pattern` is an operation that makes sense on its own. But `Pattern_head.omega_arguments` (or something similar) doesn't feel like something I'd want in the API.

Reply:
In future patches (not yet submitted to our dear upstream reviewers) we have a `val arity : head -> int` function which can be used to simplify this particular piece of code. I think it would be okay to leave this simplification for later -- when `arity` becomes useful for other reasons.

Reply:
Ok, both arguments sound good to me.